### ISSUE TITLE
Method Errors Cause Missing Field

### ISSUE THREAD

**Issue by layoaster** [marshmallow-code/marshmallow#993]
Consider the following Schema:

```python
class CallRecord(Schema):
    #: Call start timestamp.
    started_at = fields.Integer(required=True)
    #: Call answer timestamp.
    answered_at = fields.Integer(required=True, allow_none=True)
    #: Call waiting time in seconds.
    waiting = fields.Method('get_waiting_time')

    def get_waiting_time(self, record: Schema) -> int:
        answer_duration = None

        if record.answered_at:
            start_ts = datetime.fromtimestamp(record.started_at)
            answer_ts = datetime.fromtimestamp(record.answered_at)
            answer_duration = int((answer_ts - start_ts).total_seconds())

        return answer_duration
```

Consider the following data:
```python
raw_data = {
       "id": 79462182,
       "direct_link": "https://api.aircall.io/v1/calls/79462182",
       "direction": "outbound",
       "status": "done",
       "missed_call_reason": None,
       "started_at": 1539177181,
       "answered_at": 1539177194,
       "ended_at": 1539177918,
       "duration": 736
}
```

When I call `dump` you can see the `data` dictionary doesn't contain the field `waiting`.

```python
test = CallRecord().dump(raw_data)
print(test)
MarshalResult(
    data={'id': 79462182, 'answered_at': 1539177194, 'direction': 'outbound', 'duration': 736, 'status': 'done', 'started_at': 1539177181}, 
    errors={})

```

In fact the function doesn't even get called. Any idea why ?

I'm  running on Python 3.6.4 with `marshmallow==2.16.0`

**Comment by deckar01** [comment]
`record.foo` -> `record['foo']`

It looks like marshmallow 2.x is swallowing exceptions indiscriminately and ignoring the field. The issue does not seem to be present in marshmallow 3.

It should probably only catch `ValidationError`s and let the rest of them bubble up to the developer.

**Comment by layoaster** [comment]
Thanks @deckar01 Indeed, it silences the `AttributeError` exception. 

But also the [documentation](https://marshmallow.readthedocs.io/en/2.x-line/custom_fields.html#method-fields) for 2.x is misleading since it suggests that a NamedTuple-like attribute access is possible.

**Comment by deckar01** [comment]
The object provided to the method will be whatever was passed into `dump`.

**Comment by sloria** [comment]
This is a known wart in marshmallow 2. See https://github.com/marshmallow-code/marshmallow/issues/395 for the discussion. It was [fixed in 3.0](https://marshmallow.readthedocs.io/en/3.0/changelog.html#a1-2017-02-26). We don't intend to backport the fix 2.x because it may break users' existing code.
